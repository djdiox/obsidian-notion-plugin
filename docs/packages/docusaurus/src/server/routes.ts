/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import query from 'querystring';
import _ from 'lodash';
import logger from '@docusaurus/logger';
import {
  docuHash,
  normalizeUrl,
  simpleHash,
  escapePath,
} from '@docusaurus/utils';
import {getAllFinalRoutes} from './utils';
import type {
  Module,
  RouteConfig,
  RouteModules,
  ChunkNames,
  RouteChunkNames,
  ReportingSeverity,
} from '@docusaurus/types';

type LoadedRoutes = {
  /** Serialized routes config that can be directly emitted into temp file. */
  routesConfig: string;
  /** @see {ChunkNames} */
  routesChunkNames: RouteChunkNames;
  /**
   * A map from chunk name to module paths. Module paths would have backslash
   * escaped already, so they can be directly printed.
   */
  registry: {
    [chunkName: string]: string;
  };
  /**
   * Collect all page paths for injecting it later in the plugin lifecycle.
   * This is useful for plugins like sitemaps, redirects etc... Only collects
   * "actual" pages, i.e. those without subroutes, because if a route has
   * subroutes, it is probably a wrapper.
   */
  routesPaths: string[];
};

/** Indents every line of `str` by one level. */
function indent(str: string) {
  return `  ${str.replace(/\n/g, `\n  `)}`;
}

const chunkNameCache = new Map<string, string>();

/**
 * Generates a unique chunk name that can be used in the chunk registry.
 *
 * @param modulePath A path to generate chunk name from. The actual value has no
 * semantic significance.
 * @param prefix A prefix to append to the chunk name, to avoid name clash.
 * @param preferredName Chunk names default to `modulePath`, and this can supply
 * a more human-readable name.
 * @param shortId When `true`, the chunk name would only be a hash without any
 * other characters. Useful for bundle size. Defaults to `true` in production.
 */
export function genChunkName(
  modulePath: string,
  prefix?: string,
  preferredName?: string,
  shortId: boolean = process.env.NODE_ENV === 'production',
): string {
  let chunkName = chunkNameCache.get(modulePath);
  if (!chunkName) {
    if (shortId) {
      chunkName = simpleHash(modulePath, 8);
    } else {
      let str = modulePath;
      if (preferredName) {
        const shortHash = simpleHash(modulePath, 3);
        str = `${preferredName}${shortHash}`;
      }
      const name = str === '/' ? 'index' : docuHash(str);
      chunkName = prefix ? `${prefix}---${name}` : name;
    }
    chunkNameCache.set(modulePath, chunkName);
  }
  return chunkName;
}

/**
 * Takes a piece of route config, and serializes it into raw JS code. The shape
 * is the same as react-router's `RouteConfig`. Formatting is similar to
 * `JSON.stringify` but without all the quotes.
 */
function serializeRouteConfig({
  routePath,
  routeHash,
  exact,
  subroutesCodeStrings,
  props,
}: {
  routePath: string;
  routeHash: string;
  exact?: boolean;
  subroutesCodeStrings?: string[];
  props: {[propName: string]: unknown};
}) {
  const parts = [
    `path: '${routePath}'`,
    `component: ComponentCreator('${routePath}', '${routeHash}')`,
  ];

  if (exact) {
    parts.push(`exact: true`);
  }

  if (subroutesCodeStrings) {
    parts.push(
      `routes: [
${indent(subroutesCodeStrings.join(',\n'))}
]`,
    );
  }

  Object.entries(props).forEach(([propName, propValue]) => {
    // Inspired by https://github.com/armanozak/should-quote/blob/main/packages/should-quote/src/lib/should-quote.ts
    const shouldQuote = ((key: string) => {
      // Pre-sanitation to prevent injection
      if (/[.,;:}/\s]/.test(key)) {
        return true;
      }
      try {
        // If this key can be used in an expression like ({a:0}).a
        // eslint-disable-next-line no-eval
        eval(`({${key}:0}).${key}`);
        return false;
      } catch {
        return true;
      }
    })(propName);
    // Escape quotes as well
    const key = shouldQuote ? JSON.stringify(propName) : propName;
    parts.push(`${key}: ${JSON.stringify(propValue)}`);
  });

  return `{
${indent(parts.join(',\n'))}
}`;
}

const isModule = (value: unknown): value is Module =>
  typeof value === 'string' ||
  (typeof value === 'object' &&
    // eslint-disable-next-line no-underscore-dangle
    !!(value as {[key: string]: unknown} | null)?.__import);

/**
 * Takes a {@link Module} (which is nothing more than a path plus some metadata
 * like query) and returns the string path it represents.
 */
function getModulePath(target: Module): string {
  if (typeof target === 'string') {
    return target;
  }
  const queryStr = target.query ? `?${query.stringify(target.query)}` : '';
  return `${target.path}${queryStr}`;
}

/**
 * Takes a route module (which is a tree of modules), and transforms each module
 * into a chunk name. It also mutates `res.registry` and registers the loaders
 * for each chunk.
 *
 * @param routeModule One route module to be transformed.
 * @param prefix Prefix passed to {@link genChunkName}.
 * @param name Preferred name passed to {@link genChunkName}.
 * @param res The route structures being loaded.
 */
function genChunkNames(
  routeModule: RouteModules,
  prefix: string,
  name: string,
  res: LoadedRoutes,
): ChunkNames;
function genChunkNames(
  routeModule: RouteModules | RouteModules[] | Module,
  prefix: string,
  name: string,
  res: LoadedRoutes,
): ChunkNames | ChunkNames[] | string;
function genChunkNames(
  routeModule: RouteModules | RouteModules[] | Module,
  prefix: string,
  name: string,
  res: LoadedRoutes,
): string | ChunkNames | ChunkNames[] {
  if (isModule(routeModule)) {
    // This is a leaf node, no need to recurse
    const modulePath = getModulePath(routeModule);
    const chunkName = genChunkName(modulePath, prefix, name);
    res.registry[chunkName] = escapePath(modulePath);
    return chunkName;
  }
  if (Array.isArray(routeModule)) {
    return routeModule.map((val, index) =>
      genChunkNames(val, `${index}`, name, res),
    );
  }
  return _.mapValues(routeModule, (v, key) => genChunkNames(v, key, name, res));
}

export function handleDuplicateRoutes(
  pluginsRouteConfigs: RouteConfig[],
  onDuplicateRoutes: ReportingSeverity,
): void {
  if (onDuplicateRoutes === 'ignore') {
    return;
  }
  const allRoutes: string[] = getAllFinalRoutes(pluginsRouteConfigs).map(
    (routeConfig) => routeConfig.path,
  );
  const seenRoutes = new Set<string>();
  const duplicatePaths = allRoutes.filter((route) => {
    if (seenRoutes.has(route)) {
      return true;
    }
    seenRoutes.add(route);
    return false;
  });
  if (duplicatePaths.length > 0) {
    logger.report(
      onDuplicateRoutes,
    )`Duplicate routes found!${duplicatePaths.map(
      (duplicateRoute) =>
        logger.interpolate`Attempting to create page at url=${duplicateRoute}, but a page already exists at this route.`,
    )}
This could lead to non-deterministic routing behavior.`;
  }
}

/**
 * This is the higher level overview of route code generation. For each route
 * config node, it returns the node's serialized form, and mutates `registry`,
 * `routesPaths`, and `routesChunkNames` accordingly.
 */
function genRouteCode(routeConfig: RouteConfig, res: LoadedRoutes): string {
  const {
    path: routePath,
    component,
    modules = {},
    context,
    routes: subroutes,
    priority,
    exact,
    ...props
  } = routeConfig;

  if (typeof routePath !== 'string' || !component) {
    throw new Error(
      `Invalid route config: path must be a string and component is required.
${JSON.stringify(routeConfig)}`,
    );
  }

  if (!subroutes) {
    res.routesPaths.push(routePath);
  }

  const routeHash = simpleHash(JSON.stringify(routeConfig), 3);
  res.routesChunkNames[`${routePath}-${routeHash}`] = {
    // Avoid clash with a prop called "component"
    ...genChunkNames({__comp: component}, 'component', component, res),
    ...(context &&
      genChunkNames({__context: context}, 'context', routePath, res)),
    ...genChunkNames(modules, 'module', routePath, res),
  };

  return serializeRouteConfig({
    routePath: routePath.replace(/'/g, "\\'"),
    routeHash,
    subroutesCodeStrings: subroutes?.map((r) => genRouteCode(r, res)),
    exact,
    props,
  });
}

/**
 * Routes are prepared into three temp files:
 *
 * - `routesConfig`, the route config passed to react-router. This file is kept
 * minimal, because it can't be code-splitted.
 * - `routesChunkNames`, a mapping from route paths (hashed) to code-splitted
 * chunk names.
 * - `registry`, a mapping from chunk names to options for react-loadable.
 */
export function loadRoutes(
  routeConfigs: RouteConfig[],
  baseUrl: string,
  onDuplicateRoutes: ReportingSeverity,
): LoadedRoutes {
  handleDuplicateRoutes(routeConfigs, onDuplicateRoutes);
  const res: LoadedRoutes = {
    // To be written by `genRouteCode`
    routesConfig: '',
    routesChunkNames: {},
    registry: {},
    routesPaths: [normalizeUrl([baseUrl, '404.html'])],
  };

  // `genRouteCode` would mutate `res`
  const routeConfigSerialized = routeConfigs
    .map((r) => genRouteCode(r, res))
    .join(',\n');

  res.routesConfig = `import React from 'react';
import ComponentCreator from '@docusaurus/ComponentCreator';

export default [
${indent(routeConfigSerialized)},
  {
    path: '*',
    component: ComponentCreator('*'),
  },
];
`;

  return res;
}
