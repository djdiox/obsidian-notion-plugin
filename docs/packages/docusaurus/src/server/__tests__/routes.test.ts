/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {jest} from '@jest/globals';
import {loadRoutes, handleDuplicateRoutes, genChunkName} from '../routes';
import type {RouteConfig} from '@docusaurus/types';

describe('genChunkName', () => {
  it('works', () => {
    const firstAssert: {[key: string]: string} = {
      '/docs/adding-blog': 'docs-adding-blog-062',
      '/docs/versioning': 'docs-versioning-8a8',
      '/': 'index',
      '/blog/2018/04/30/How-I-Converted-Profilo-To-Docusaurus':
        'blog-2018-04-30-how-i-converted-profilo-to-docusaurus-4f2',
      '/youtube': 'youtube-429',
      '/users/en/': 'users-en-f7a',
      '/blog': 'blog-c06',
    };
    Object.keys(firstAssert).forEach((str) => {
      expect(genChunkName(str)).toBe(firstAssert[str]);
    });
  });

  it("doesn't allow different chunk name for same path", () => {
    expect(genChunkName('path/is/similar', 'oldPrefix')).toEqual(
      genChunkName('path/is/similar', 'newPrefix'),
    );
  });

  it('emits different chunk names for different paths even with same preferred name', () => {
    const secondAssert: {[key: string]: string} = {
      '/blog/1': 'blog-85-f-089',
      '/blog/2': 'blog-353-489',
    };
    Object.keys(secondAssert).forEach((str) => {
      expect(genChunkName(str, undefined, 'blog')).toBe(secondAssert[str]);
    });
  });

  it('only generates short unique IDs', () => {
    const thirdAssert: {[key: string]: string} = {
      a: '0cc175b9',
      b: '92eb5ffe',
      c: '4a8a08f0',
      d: '8277e091',
    };
    Object.keys(thirdAssert).forEach((str) => {
      expect(genChunkName(str, undefined, undefined, true)).toBe(
        thirdAssert[str],
      );
    });
    expect(genChunkName('d', undefined, undefined, true)).toBe('8277e091');
  });
});

describe('handleDuplicateRoutes', () => {
  const routes: RouteConfig[] = [
    {
      path: '/',
      component: '',
      routes: [
        {path: '/search', component: ''},
        {path: '/sameDoc', component: ''},
      ],
    },
    {
      path: '/',
      component: '',
      routes: [
        {path: '/search', component: ''},
        {path: '/sameDoc', component: ''},
        {path: '/uniqueDoc', component: ''},
      ],
    },
    {
      path: '/',
      component: '',
    },
    {
      path: '/',
      component: '',
    },
    {
      path: '/',
      component: '',
    },
  ];
  it('works', () => {
    expect(() => {
      handleDuplicateRoutes(routes, 'throw');
    }).toThrowErrorMatchingSnapshot();
    const consoleMock = jest.spyOn(console, 'log').mockImplementation(() => {});
    handleDuplicateRoutes(routes, 'ignore');
    expect(consoleMock).toBeCalledTimes(0);
  });
});

describe('loadRoutes', () => {
  it('loads nested route config', () => {
    const nestedRouteConfig: RouteConfig = {
      component: '@theme/DocPage',
      path: '/docs:route',
      modules: {
        docsMetadata: 'docs-b5f.json',
      },
      routes: [
        {
          path: '/docs/hello',
          component: '@theme/DocItem',
          exact: true,
          modules: {
            content: 'docs/hello.md',
            metadata: 'docs-hello-da2.json',
          },
          context: {
            plugin: 'pluginRouteContextModule-100.json',
          },
          sidebar: 'main',
        },
        {
          path: 'docs/foo/baz',
          component: '@theme/DocItem',
          modules: {
            content: 'docs/foo/baz.md',
            metadata: 'docs-foo-baz-dd9.json',
          },
          context: {
            plugin: 'pluginRouteContextModule-100.json',
          },
          sidebar: 'secondary',
          'key:a': 'containing colon',
          "key'b": 'containing quote',
          'key"c': 'containing double quote',
          'key,d': 'containing comma',
          字段: 'containing unicode',
        },
      ],
    };
    expect(loadRoutes([nestedRouteConfig], '/', 'ignore')).toMatchSnapshot();
  });

  it('loads flat route config', () => {
    const flatRouteConfig: RouteConfig = {
      path: '/blog',
      component: '@theme/BlogListPage',
      exact: true,
      modules: {
        items: [
          {
            content: {
              __import: true,
              path: 'blog/2018-12-14-Happy-First-Birthday-Slash.md',
              query: {
                truncated: true,
              },
            },
            metadata: 'blog-2018-12-14-happy-first-birthday-slash-d2c.json',
          },
          {
            content: 'blog/2018-12-14-Happy-First-Birthday-Slash.md',
          },
          {
            content: {
              __import: true,
              path: 'blog/2018-12-14-Happy-First-Birthday-Slash.md',
            },
          },
        ],
      },
    };
    expect(loadRoutes([flatRouteConfig], '/', 'ignore')).toMatchSnapshot();
  });

  it('rejects invalid route config', () => {
    const routeConfigWithoutPath = {
      component: 'hello/world.js',
    } as RouteConfig;

    expect(() => loadRoutes([routeConfigWithoutPath], '/', 'ignore'))
      .toThrowErrorMatchingInlineSnapshot(`
      "Invalid route config: path must be a string and component is required.
      {"component":"hello/world.js"}"
    `);

    const routeConfigWithoutComponent = {
      path: '/hello/world',
    } as RouteConfig;

    expect(() => loadRoutes([routeConfigWithoutComponent], '/', 'ignore'))
      .toThrowErrorMatchingInlineSnapshot(`
      "Invalid route config: path must be a string and component is required.
      {"path":"/hello/world"}"
    `);
  });

  it('loads route config with empty (but valid) path string', () => {
    const routeConfig = {
      path: '',
      component: 'hello/world.js',
    } as RouteConfig;

    expect(loadRoutes([routeConfig], '/', 'ignore')).toMatchSnapshot();
  });
});
