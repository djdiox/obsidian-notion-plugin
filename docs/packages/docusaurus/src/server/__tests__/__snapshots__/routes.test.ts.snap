// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleDuplicateRoutes works 1`] = `
"Duplicate routes found!
- Attempting to create page at /search, but a page already exists at this route.
- Attempting to create page at /sameDoc, but a page already exists at this route.
- Attempting to create page at /, but a page already exists at this route.
- Attempting to create page at /, but a page already exists at this route.
This could lead to non-deterministic routing behavior."
`;

exports[`loadRoutes loads flat route config 1`] = `
{
  "registry": {
    "__comp---theme-blog-list-pagea-6-a-7ba": "@theme/BlogListPage",
    "content---blog-0-b-4-09e": "blog/2018-12-14-Happy-First-Birthday-Slash.md?truncated=true",
    "content---blog-7-b-8-fd9": "blog/2018-12-14-Happy-First-Birthday-Slash.md",
    "metadata---blog-0-b-6-74c": "blog-2018-12-14-happy-first-birthday-slash-d2c.json",
  },
  "routesChunkNames": {
    "/blog-599": {
      "__comp": "__comp---theme-blog-list-pagea-6-a-7ba",
      "items": [
        {
          "content": "content---blog-0-b-4-09e",
          "metadata": "metadata---blog-0-b-6-74c",
        },
        {
          "content": "content---blog-7-b-8-fd9",
        },
        {
          "content": "content---blog-7-b-8-fd9",
        },
      ],
    },
  },
  "routesConfig": "import React from 'react';
import ComponentCreator from '@docusaurus/ComponentCreator';

export default [
  {
    path: '/blog',
    component: ComponentCreator('/blog', '599'),
    exact: true
  },
  {
    path: '*',
    component: ComponentCreator('*'),
  },
];
",
  "routesPaths": [
    "/404.html",
    "/blog",
  ],
}
`;

exports[`loadRoutes loads nested route config 1`] = `
{
  "registry": {
    "__comp---theme-doc-item-178-a40": "@theme/DocItem",
    "__comp---theme-doc-page-1-be-9be": "@theme/DocPage",
    "content---docs-foo-baz-8-ce-61e": "docs/foo/baz.md",
    "content---docs-helloaff-811": "docs/hello.md",
    "docsMetadata---docs-routef-34-881": "docs-b5f.json",
    "metadata---docs-foo-baz-2-cf-fa7": "docs-foo-baz-dd9.json",
    "metadata---docs-hello-956-741": "docs-hello-da2.json",
    "plugin---docs-hello-665-3ca": "pluginRouteContextModule-100.json",
  },
  "routesChunkNames": {
    "/docs/hello-fcc": {
      "__comp": "__comp---theme-doc-item-178-a40",
      "__context": {
        "plugin": "plugin---docs-hello-665-3ca",
      },
      "content": "content---docs-helloaff-811",
      "metadata": "metadata---docs-hello-956-741",
    },
    "/docs:route-502": {
      "__comp": "__comp---theme-doc-page-1-be-9be",
      "docsMetadata": "docsMetadata---docs-routef-34-881",
    },
    "docs/foo/baz-eb2": {
      "__comp": "__comp---theme-doc-item-178-a40",
      "__context": {
        "plugin": "plugin---docs-hello-665-3ca",
      },
      "content": "content---docs-foo-baz-8-ce-61e",
      "metadata": "metadata---docs-foo-baz-2-cf-fa7",
    },
  },
  "routesConfig": "import React from 'react';
import ComponentCreator from '@docusaurus/ComponentCreator';

export default [
  {
    path: '/docs:route',
    component: ComponentCreator('/docs:route', '502'),
    routes: [
      {
        path: '/docs/hello',
        component: ComponentCreator('/docs/hello', 'fcc'),
        exact: true,
        sidebar: "main"
      },
      {
        path: 'docs/foo/baz',
        component: ComponentCreator('docs/foo/baz', 'eb2'),
        sidebar: "secondary",
        "key:a": "containing colon",
        "key'b": "containing quote",
        "key\\"c": "containing double quote",
        "key,d": "containing comma",
        字段: "containing unicode"
      }
    ]
  },
  {
    path: '*',
    component: ComponentCreator('*'),
  },
];
",
  "routesPaths": [
    "/404.html",
    "/docs/hello",
    "docs/foo/baz",
  ],
}
`;

exports[`loadRoutes loads route config with empty (but valid) path string 1`] = `
{
  "registry": {
    "__comp---hello-world-jse-0-f-b6c": "hello/world.js",
  },
  "routesChunkNames": {
    "-b2a": {
      "__comp": "__comp---hello-world-jse-0-f-b6c",
    },
  },
  "routesConfig": "import React from 'react';
import ComponentCreator from '@docusaurus/ComponentCreator';

export default [
  {
    path: '',
    component: ComponentCreator('', 'b2a')
  },
  {
    path: '*',
    component: ComponentCreator('*'),
  },
];
",
  "routesPaths": [
    "/404.html",
    "",
  ],
}
`;
